### ISSUE TITLE
Support for anything-but/wildcard

### ISSUE THREAD

**Issue by schenksj** [aws/event-ruler#72]
## What is your idea?
Support wildcard matches in anything-but:

```
{
  "detail": {
    "command_path": [ { "anything-but": {"wildcard": "*cmd.exe*mycommand*"} } ]
  }
}
```

## Would you be willing to make the change? 
Probably

## Additional context

Add any other context (such as images, docs, posts) about the idea here.

**Comment by baldawar** [comment]
Gotta be honest, was expect this to pop up eventually. 

I don't have objections to the feature itself but there are performance implications of using wildcards, so benchmarks need to be thorough. Wildcards have also not been validated against production traffic (or at least nobody has shared it with me), so I'd encourage additional testing within the library but also within your end application.

We also have a new effort with this class that helps reason complexity tied to wildcards https://github.com/aws/event-ruler/blob/main/src/main/software/amazon/event/ruler/MachineComplexityEvaluator.java, but it won't work because of line 178.

**Comment by schenksj** [comment]
Thanks @baldawar !  I'll probably take a look at this at some point after #71 is done, its not terribly high priority for me but it only looks like a few minutes of work.  Quick question, is my assumption that wildcard matching only accepts one pattern (not a list) as input correct?

**Comment by baldawar** [comment]
I'm expecting this has some trickiness. Wildcard involves compound byte transitions and opens risk of forever looping. The MachineComplextyEvaluator is also an unknown in if it'll work. 

In general, wildcard performance can downgrade performance (see https://github.com/aws/event-ruler#matching-with-a-machine for some numbers), so I'm wondering if it'll be worth the effort.

**Comment by schenksj** [comment]
@baldawar I think for now anything-but enhancements are lower priority for me.  I came to terms with the limitations and in my app have a secondary "if-matches-exclude" rule capability that can be applied on top of rules.  The bigger annoyance for now is around case-insensitivity in wildcards (and suffix/prefix though you can accomplish similar work with wildcards).  My work-around is less elegant (pre-processing to re-present fields in lowercase before running matches).

I'm comfortable shelving this one for now.  Thank you for your thoughts.

**Comment by baldawar** [comment]
I'm going to leave the issues open for a while in case it someone else wants to dig into them later.  

on 

>  The bigger annoyance for now is around case-insensitivity in wildcards (and suffix/prefix though you can accomplish similar work with wildcards). My work-around is less elegant (pre-processing to re-present fields in lowercase before running matches).

To `equals-ignore-case` works only on the leaf nodes. If it was made to work with subsequent matchers similar to anything-but, like `equals-ignore-case: {"anything-but": {"prefix": "abc"}`, it would solve your problem. I haven't looked into the matcher for ignores-case for a while, but I still suspect its a lot more work than just moving normal latin characters to lower-case.
